File: Android/core/src/main/java/com/khushi/conduit/core/link/HotspotRequest.kt
package com.khushi.conduit.core.link

import com.khushi.conduit.core.protocol.LinkCrypto
import java.nio.ByteBuffer
import kotlin.math.abs

/**
 * A Mac asking, over Bluetooth LE, for this phone's hotspot —
 * Shared/Protocol/README.md §8a. The Mac signs it with its Conduit Link
 * identity key and addresses it to one phone.
 */
object HotspotRequest {

    const val SERVICE_UUID = "9D99F667-22A5-4207-B9F0-CF2A1F80D81B"
    const val CHARACTERISTIC_UUID = "A9FE1931-A845-4B54-811D-9095A568370A"
    private const val VERSION: Byte = 1
    private const val HEADER = 1 + 16 + 16 + 8 + 16

    class Fields(val macId: ByteArray, val phoneId: ByteArray, val timestamp: Long, val nonce: ByteArray)

    /** `"conduit-hotspot-v1" ‖ macID ‖ phoneID ‖ timestamp ‖ nonce` */
    fun statement(fields: Fields): ByteArray =
        "conduit-hotspot-v1".toByteArray() + fields.macId + fields.phoneId +
            ByteBuffer.allocate(8).putLong(fields.timestamp).array() + fields.nonce

    /** The fields and the signature, or null when the bytes are not a request. */
    fun decode(bytes: ByteArray): Pair<Fields, ByteArray>? {
        if (bytes.size <= HEADER || bytes[0] != VERSION) return null
        val fields = Fields(
            macId = bytes.copyOfRange(1, 17),
            phoneId = bytes.copyOfRange(17, 33),
            timestamp = ByteBuffer.wrap(bytes, 33, 8).long,
            nonce = bytes.copyOfRange(41, 57),
        )
        return fields to bytes.copyOfRange(HEADER, bytes.size)
    }

    fun hex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }
}

/**
 * Decides whether a hotspot request deserves a notification. It must come
 * from a paired Mac, be addressed to this phone, carry that Mac's signature,
 * and be fresh: within two minutes of this phone's clock, with a nonce not
 * seen in the last five. Anything else is dropped without an answer.
 */
class HotspotRequestGate(private val now: () -> Long = System::currentTimeMillis) {

    private val seen = LinkedHashMap<String, Long>()

    /** The device ID of the Mac that asked, or null to ignore the request. */
    @Synchronized
    fun check(request: ByteArray, phoneId: String, pairedMacKeys: Map<String, ByteArray>): String? {
        val (fields, signature) = HotspotRequest.decode(request) ?: return null
        val macId = HotspotRequest.hex(fields.macId)
        val macKey = pairedMacKeys[macId] ?: return null
        if (HotspotRequest.hex(fields.phoneId) != phoneId) return null

        val time = now()
        if (abs(time - fields.timestamp) > FRESH_MILLIS) return null
        seen.entries.removeAll { time - it.value > NONCE_MEMORY_MILLIS }
        val nonce = HotspotRequest.hex(fields.nonce)
        if (nonce in seen) return null

        if (!LinkCrypto.verify(signature, HotspotRequest.statement(fields), macKey)) return null
        seen[nonce] = time
        return macId
    }

    private companion object {
        const val FRESH_MILLIS = 2 * 60_000L
        const val NONCE_MEMORY_MILLIS = 5 * 60_000L
    }
}

File: Shared/ConduitKit/Sources/ConduitCore/Link/HotspotRequest.swift
//
//  HotspotRequest.swift
//  ConduitCore
//
//  The request this Mac sends over Bluetooth LE when it wants the phone's
//  hotspot — Shared/Protocol/README.md §8a. When the Mac has no network it
//  cannot reach the phone over Conduit Link, so it signs a short request with
//  its identity key instead, addressed to one paired phone.
//

import CryptoKit
import Foundation

nonisolated enum HotspotRequest {

    static let serviceUUID = "9D99F667-22A5-4207-B9F0-CF2A1F80D81B"
    static let characteristicUUID = "A9FE1931-A845-4B54-811D-9095A568370A"
    static let version: UInt8 = 1

    struct Fields: Equatable, Sendable {
        /// Raw 16-byte device IDs (§4).
        let macID: Data
        let phoneID: Data
        /// Milliseconds since the Unix epoch.
        let timestamp: UInt64
        let nonce: Data
    }

    /// `"conduit-hotspot-v1" ‖ macID ‖ phoneID ‖ timestamp ‖ nonce`
    static func statement(_ fields: Fields) -> Data {
        Data("conduit-hotspot-v1".utf8) + fields.macID + fields.phoneID + bigEndian(fields.timestamp) + fields.nonce
    }

    static func encode(_ fields: Fields, signature: Data) -> Data {
        Data([version]) + fields.macID + fields.phoneID + bigEndian(fields.timestamp) + fields.nonce + signature
    }

    /// A signed request from this Mac to one paired phone.
    static func make(identity: LinkIdentity, phoneID: String, now: Date = Date()) throws -> Data {
        guard let macID = deviceIDBytes(identity.deviceID), let phone = deviceIDBytes(phoneID) else {
            throw LinkCryptoError.badKey
        }
        let fields = Fields(macID: macID, phoneID: phone,
                            timestamp: UInt64(now.timeIntervalSince1970 * 1000),
                            nonce: Data((0 ..< 16).map { _ in UInt8.random(in: .min ... .max) }))
        return encode(fields, signature: try LinkCrypto.sign(statement(fields), with: identity.signingKey))
    }

    /// The fields and signature of a request, or nil if it is not one.
    static func decode(_ data: Data) -> (fields: Fields, signature: Data)? {
        let header = 1 + 16 + 16 + 8 + 16
        guard data.count > header, data.first == version else { return nil }
        let bytes = [UInt8](data)
        let timestamp = bytes[33 ..< 41].reduce(UInt64(0)) { ($0 << 8) | UInt64($1) }
        return (Fields(macID: Data(bytes[1 ..< 17]), phoneID: Data(bytes[17 ..< 33]),
                       timestamp: timestamp, nonce: Data(bytes[41 ..< 57])),
                Data(bytes[header...]))
    }

    /// A device ID's 32 hex characters as its 16 raw bytes.
    static func deviceIDBytes(_ id: String) -> Data? {
        guard id.count == 32 else { return nil }
        var bytes = Data(capacity: 16)
        var index = id.startIndex
        while index < id.endIndex {
            let next = id.index(index, offsetBy: 2)
            guard let byte = UInt8(id[index ..< next], radix: 16) else { return nil }
            bytes.append(byte)
            index = next
        }
        return bytes
    }

    private static func bigEndian(_ value: UInt64) -> Data {
        withUnsafeBytes(of: value.bigEndian) { Data($0) }
    }
}
